Extract encoding logic into a separate class

camertron opened the pull request:
_**This pull request is part of a larger effort to bring OTF support to TTFunk. See https://github.com/prawnpdf/ttfunk/issues/53 for details.**_

Higlights of this PR:
1. Currently font encoding (i.e. serializing the font into a series of bytes that can be written to a file) is performed in the `Subset::Base` class. This PR extracts it into its own class.
2. Introduce the `EncodedString` class capable of aggregating encoded bytes and tracking placeholder locations.

camertron commented on lib/ttfunk/encoded_string.rb:
```diff
@@ -0,0 +1,58 @@
+      io.string
+    end
+
+    def add_placeholder(category, name, position: nil, length: 1)
```
Ok, I like the padding idea.

It's so funny you would say that `<<` should accept placeholders as well as raw bytes. The original version of this class accepts a number of different types for `<<` but I wanted to make it simpler for the purposes of this PR. I have a bunch of code written for the GSUB and GPOS tables, both of which require extensive use of relative and non-relative placeholders. Since those tables will come (much) later, I didn't think it made sense to complicate `<<`. However, if you like the idea of allowing it to accept `Placeholder` objects, then fine by me!

It's worth noting that `add_placeholder` does not require `position` to be passed and falls back to `io.pos` if none is provided (or it's `nil`). I think I'll just get rid of `add_placeholder` and rely entirely on `<<`.

I originally thought both `name` and `category` were necessary for my eventual implementation of the GPOS and GSUB tables, however in thinking about it a little more I think we can probably get by with just `name`.

There can be multiple placeholders with the same name to support the use case where you want to reference the same data from multiple locations, such as the coverage tables in GPOS/GSUB. Even though supporting multiple locations isn't necessary for this PR, it doesn't introduce too much complexity to support.

pointlessone replied:
I might have misunderstood the purpose of this class and thus have incorrect metal model. Consequently, I may propose inadequate changes so please feel free to correct me.

I see this class as a partially serialized structure. It gives a series of bytes for a proper data structure that can differ in just a few places when we change a field in the data structure. It's an optimization that saves us a few CPU cycles because we don't have to sully serialize the structure when only one field changes.

For instance, you use it to populate table offsets in TTF directory. Technically, you can just calculate those offsets as you have all the data necessary at hand but that's OK. It works.

A good use for a placeholder would be checksum calculation. You need the whole font serialized with `0` checksum before you can get a real checksum. Then you can properly serialize he whole font. But you don't have to as only the checksum is different. You replace only those bytes and you're done. BTW, you've chosen to manually rewrite those bytes even though you have placeholders implemented. 🤷‍♂️ 

I see placeholders as named addresses in the stream. And the whole thing is only useful as an optimization. Effectively the same result can be achieved by double serialization of the source tables.

With regards of multiple placeholder positions. I can think only of one case when they might be useful: when a value is used in multiple places in the structure but is not readily available. I'm not quite an expert on OTF (or any other SFNT-based format for that matter) but I read specs and worked a bit with the format. I can't recall a single instance of value reuse. Please point me to a relevant places in the spec if you're aware of any.

So… Multiple positions seem to be out of scope in this PR for a few reasons:
* First of all it's unnecessary right now
* From a cursory inspection of the OTF spec for [Coverage table](https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2#coverage-table) it seems there's no value reuse so I'm not sure how multiple positions can be useful
* #53 doesn't even mention GPOS/GSUB tables. I only found those mentioned once in the source code noting that they're not implemented.

camertron replied:
> I see this class as a partially serialized structure. It gives a series of bytes for a proper data structure that can differ in just a few places when we change a field in the data structure. It's an optimization that saves us a few CPU cycles because we don't have to fully serialize the structure when only one field changes.

Yes, that's my mental model as well.

> A good use for a placeholder would be checksum calculation. BTW, you've chosen to manually rewrite those bytes even though you have placeholders implemented. 🤷‍♂️

That was not a choice I personally made, it was present in the original code. It's a good opportunity for refactoring however :)

> So… Multiple positions seem to be out of scope in this PR for a few reasons:

* Perhaps unnecessary now, but as I mentioned earlier multiple values per name is not difficult or complicated to support. Basically I see no downside.
* I looked through the coverage table spec at length looking for any mention of reuse and didn't find anything. I can however say from empirically gathered evidence that coverage tables are reused between lookup subtables. If you're curious, check out my `everything` branch and run `bundle console`. This snippet of code illustrates the reuse:
```ruby
font = TTFunk::File.open('/path/to/NotoSerifDisplay-Regular.ttf')
groups = font.glyph_substitutions.lookup_list.tables.flat_map do |tb|
  tb.sub_tables.map do |sub|
    sub.respond_to?(:coverage_table) ? sub.coverage_table : nil
  end.compact
end.each_with_object({}) do |cov, ret|
  ret[cov.table_offset] ||= []
  ret[cov.table_offset] << cov
end

groups.map { |_, covs| covs.size }  # [1, 1, 2, 2, 1, ..., 3, ..., 1, 1]
```
* That's correct, #53 does not mention the GPOS/GSUB tables. This is because they are not actually necessary for supporting Open-Type fonts (I initially thought they were). I'm hoping to add support for them later on as a separate PR (or series of PRs).

pointlessone replied:
> not difficult or complicated to support. Basically I see no downside.

There's a downside of maintaining code that serves no purpose. If for any reason you won't get to actually using this feature and won't be around to answer questions future contributors/maintainers will get confused. This is not a nice thing to have.

I strongly lean towards not implementing potentially useful features until they are actually used.

I understand that you envision this to be used pretty soon :tm: but unless you can guarantee delivering in a reasonable timeframe I'd rather not have this in the codebase.

Just to be clear, I don't need the guarantee. That's not how Open Source generally work. I tremendously appreciate your effort here but don't want to introduce decoherence (at least more than there's already) to the codebase. Please try keeping amount of code to the minimum. Refactor when needed.

Let's keep this out for now. I won't say a word when you'll actually need it.

I also look forward to GPOS/GSUB tables implemented but those are quite removed in time.

> coverage tables are reused between lookup subtables

It very well might be. However, it doesn't look like there's a reuse opportunity within current architecture. We have tables serialized independently. It doesn't seem there's any cross-table reuse opportunity without higher complexity involved. At least at the first glance.

> That was not a choice I personally made, it was present in the original code. It's a good opportunity for refactoring however :)

Sorry, I assumed you took the opportunity since you changed one part of the code to use `EncodedString`.

Yes, I think you should use placeholders for checksum as well.

camertron replied:
Alright, I understand your point of view 😄 I just pushed fairly extensive changes to `EncodedString` that have removed the double placeholder stuff and added the ability to shovel on other instances of `EncodedString`. Placeholder positions are updated automatically. To make things more uniform, I also updated the `encode` methods of all the individual table classes to return instances of `EncodedString`. Now it's much easier to add a placeholder in one class (i.e. `Head`) and resolve it in the caller (i.e. `TTFEncoder`). Let me know what you think!

camertron commented on lib/ttfunk/encoded_string.rb:
```diff
@@ -0,0 +1,58 @@
+
+    def io
+      @io ||= StringIO.new.tap do |sio|
+        sio.set_encoding(::Encoding::ASCII_8BIT)
```
Nice, I thought `#binmode` was a no-op on `StringIO` :)

camertron commented on lib/ttfunk/encoded_string.rb:
```diff
@@ -0,0 +1,58 @@
+    end
+
+    def string
+      io.string
```
Sure, I like that idea.

pointlessone commented on lib/ttfunk/encoded_string.rb:
```diff
@@ -0,0 +1,97 @@
+      string.bytes
+    end
+
+    def unresolved_string
```
This and all the following public methods can be marked protected. They still will be accessible from withing this class but won't be a part of a public API.

camertron replied:
The `#unresolved_string` method is called several times outside the context of the `EncodedString` class, so it will need to remain public.

pointlessone commented on lib/ttfunk/table/cmap.rb:
```diff
@@ -8,7 +8,10 @@ def self.encode(charmap, encoding)
 
         # pack 'version' and 'table-count'
-        result[:table] = [0, 1, result.delete(:subtable)].pack('nnA*')
+        result[:table] = EncodedString.new do |table|
```
The block seem to be superfluous here since `<<` returns `self`. Moreover, I don't see how use of `EncodedString` helps here. We have a string and it seems perfectly fine in this case (and all the others like this).

pointlessone commented on lib/ttfunk/table/head.rb:
```diff
@@ -22,10 +22,18 @@ class Head < TTFunk::Table
+        EncodedString.new do |table|
+          table << [head.version, head.font_revision].pack('N2')
+          table << Placeholder.new(:checksum, length: 4)
+          table << [
```
I think, it would be OK to just chain shovels here.

```ruby
EncodedString.new << 
  [head.version, head.font_revision].pack('N2') <<
  Placeholder.new(:checksum, length: 4) <<
  [...].pack('Nn2q2n*')
```

camertron commented:
@pointlessone No problem, I too am traveling. As a matter of fact, I'm writing this from Reykjavík :)

I had hoped to make things consistent by returning instances of `EncodedString` everywhere, since it's a little strange to me to return both strings and `EncodedString`s. However, I understand your concerns. In most cases, a string will do just fine, so I have reverted the use of `EncodedString` in a number of places and fixed the tests.
